L1Trigger/L1THGCal/src/be_algorithms/HGCalMulticlusteringImpl.cc

```diff
@@ -109,10 +110,50 @@ void HGCalMulticlusteringImpl::clusterizeDR( const edm::PtrVector<l1t::HGCalClus
 
     /* making the collection of multiclusters */
     for( unsigned i(0); i<multiclustersTmp.size(); ++i ){
-        math::PtEtaPhiMLorentzVector calibP4(  multiclustersTmp.at(i).pt() * calibSF_, 
-                                               multiclustersTmp.at(i).eta(), 
-                                               multiclustersTmp.at(i).phi(), 
+
+        double calibPt=0.;
+        if(applyLayerWeights_){
+            const edm::PtrVector<l1t::HGCalCluster> &pertinentClu = multiclustersTmp.at(i).constituents();
+            for( edm::PtrVector<l1t::HGCalCluster>::const_iterator  it_clu=pertinentClu.begin(); it_clu<pertinentClu.end(); it_clu++){
+                int layerN = -1; 
+                if( (*it_clu)->subdetId()==HGCEE ){
+                    layerN = (*it_clu)->layer();
+                }
+                else if((*it_clu)->subdetId()==HGCHEF){
+                    layerN = (*it_clu)->layer()+kLayersEE_;
+                }
+                else if((*it_clu)->subdetId()==HGCHEB){
+                    layerN = (*it_clu)->layer()+kLayersFH_+kLayersEE_;
+                }
+                calibPt += layerWeights_.at(layerN) * (*it_clu)->mipPt();
+            }     
+        }
+        else{        
+            calibPt = multiclustersTmp.at(i).pt() * calibSF_; 
+        }
+
+        // compute the eta, phi observables for multicluster starting from its barycenter x,y,z position 
+        double mcl_x = multiclustersTmp.at(i).centre().x();
+        double mcl_y = multiclustersTmp.at(i).centre().y();
+        double mcl_z = multiclustersTmp.at(i).centre().z();
+
+        double mcl_rho = sqrt( mcl_x*mcl_x + mcl_y*mcl_y);
+        double cotanTheta = ( mcl_z / mcl_rho );
+
+        double mcl_eta = asinh(cotanTheta);
+
+        double mcl_phi = 0.;
+        if(mcl_y>0){
+            mcl_phi = acos( mcl_x / mcl_rho );
+        }else{
+            mcl_phi = -acos( mcl_x / mcl_rho );
+        }
+
+        math::PtEtaPhiMLorentzVector calibP4(  calibPt, 
+                                               mcl_eta, 
+                                               mcl_phi, 
```

Comment:
Use the `eta()` and `phi()` method of the `GlobalPoint` instead of recomputing them. Much less error-prone, and also more readable.